server/api.js

```diff
@@ -0,0 +1,551 @@
+
+async function handleGetClientGateway(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const gateway = await db.getGatewayWithId(req.params.gatewayId)
+
+      // check that this gateway exists and is owned by the specified client
+      // NOTE: if clientId is invalid, then the query will fail and return null
+      if (gateway == null || gateway.client.id !== req.params.clientId) {
+        res.status(404).send({ status: 'error', message: 'Not Found' })
+
+        return
+      }
+
+      res.status(200).send({ status: 'success', data: gateway })
+    } else {
+      res.status(400).send({ status: 'error', message: 'Bad Request' })
+      helpers.logError(`Bad request to ${req.path}: ${validationErrors.array()}`)
+    }
+  } catch (error) {
+    res.status(500).send({ status: 'error', message: 'Internal Server Error' })
+    helpers.logError(`Internal server error at ${req.path}: ${error.message}`)
+  }
+}
+
+const validateGetClientGateways = [Validator.header(['Authorization']).notEmpty(), Validator.param(['clientId']).notEmpty()]
+
+async function handleGetClientGateways(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const gateways = await db.getGatewaysWithClientId(req.params.clientId)
+
+      // if the query failed and returned null, the clientId is probably wrong
+      if (gateways == null) {
+        res.status(404).send({ status: 'error', message: 'Not Found' })
+
+        return
+      }
+
+      res.status(200).send({ status: 'success', data: gateways })
+    } else {
+      res.status(400).send({ status: 'error', message: 'Bad Request' })
+      helpers.logError(`Bad request to ${req.path}: ${validationErrors.array()}`)
+    }
+  } catch (error) {
+    res.status(500).send({ status: 'error', message: 'Internal Server Error' })
+    helpers.logError(`Internal server error at ${req.path}: ${error.message}`)
+  }
+}
+
+const validateGetClientVitals = [Validator.header(['Authorization']).notEmpty(), Validator.param(['clientId']).notEmpty()]
+
+async function handleGetClientVitals(req, res) {
+  const validationErrors = Validator.validationResult(req).formatWith(helpers.formatExpressValidationErrors)
+
+  try {
+    if (validationErrors.isEmpty()) {
+      const buttonVitals = await db.getRecentButtonsVitalsWithClientId(req.params.clientId)
+      const gatewayVitals = await db.getRecentGatewaysVitalsWithClientId(req.params.clientId)
+
+      // if either of the query failed and returned null, the clientId is probably wrong
+      if (buttonVitals == null || gatewayVitals == null) {
+        res.status(404).send({ status: 'error', message: 'Not Found' })
```

Comment:
It is kind of weird for an empty array to be returned if the input client ID was wrong, for example. I'll change the database function to return null in the event of an error, instead of an empty array!

Reply:
I changed my mind on this; I'm not entirely sure where these DB functions are used, so it might be better to have the safeguard of always returning an array, and returning an empty array even if an error occurred. It makes sense logically too---if a client doesn't exist, then they can't have any buttons; therefore an empty array makes sense. I've reverted my changes to these functions and changed my new functions to follow from that.

I think I'm going to try and have all API routes return 200, 201 if OK; 400 if validation fails; and 500 if an error was thrown. I'm going to rework the existing 404s to either return an empty array, null, or throw an error.